// @webex/plugin-meetings/dist/webex.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

require("@webex/plugin-authorization");

require("@webex/internal-plugin-device");

require("@webex/internal-plugin-support");

require("@webex/plugin-logger");

require("@webex/plugin-people");

require("@webex/plugin-rooms");

var _webexCore = _interopRequireWildcard(require("@webex/webex-core"));

var _merge = _interopRequireDefault(require("lodash/merge"));

var Webex = _webexCore.default.extend({
  webex: true,
  version: "1.99.0"
});

Webex.version = "1.99.0";

Webex.init = function init() {
  var attrs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var config = (0, _merge.default)({}, {
    hydra: process.env.HYDRA_SERVICE_URL || 'https://api.ciscospark.com/v1',
    credentials: {
      clientType: 'confidential'
    },
    device: {
      ephemeral: true
    },
    storage: {
      boundedAdapter: _webexCore.MemoryStoreAdapter,
      unboundedAdapter: _webexCore.MemoryStoreAdapter
    }
  }, attrs.config);
  return new Webex((0, _merge.default)({}, {
    config: config
  }, attrs));
};

var _default = Webex;
exports.default = _default;
