// @webex/plugin-meetings/dist/reconnection-manager/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs2/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/asyncToGenerator"));

var _promise = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/promise"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _construct = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/reflect/construct"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/wrapNativeSuper"));

var _loggerProxy = _interopRequireDefault(require("../common/logs/logger-proxy"));

var _triggerProxy = _interopRequireDefault(require("../common/events/trigger-proxy"));

var _constants = require("../constants");

var _reconnection = _interopRequireDefault(require("../common/errors/reconnection"));

var _reconnectionInProgress = _interopRequireDefault(require("../common/errors/reconnection-in-progress"));

var _peerConnectionManager = _interopRequireDefault(require("../peer-connection-manager"));

var _config = require("../metrics/config");

var _media = _interopRequireDefault(require("../media"));

var _metrics = _interopRequireDefault(require("../metrics"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = (0, _construct.default)(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !_construct.default) return false; if (_construct.default.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call((0, _construct.default)(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * Used to indicate that the reconnect logic needs to be retried.
 *
 * @class NeedsRetryError
 * @extends {Error}
 */
var NeedsRetryError = /*#__PURE__*/function (_Error) {
  (0, _inherits2.default)(NeedsRetryError, _Error);

  var _super = _createSuper(NeedsRetryError);

  function NeedsRetryError() {
    (0, _classCallCheck2.default)(this, NeedsRetryError);
    return _super.apply(this, arguments);
  }

  return NeedsRetryError;
}( /*#__PURE__*/(0, _wrapNativeSuper2.default)(Error));
/**
 * @export
 * @class ReconnectionManager
*/


var ReconnectionManager = /*#__PURE__*/function () {
  /**
   * @param {Meeting} meeting
   */
  function ReconnectionManager(meeting) {
    (0, _classCallCheck2.default)(this, ReconnectionManager);

    /**
     * Stores ICE reconnection state data.
     *
     * @instance
     * @type {Object}
     * @private
     * @memberof ReconnectionManager
     */
    this.iceState = {
      disconnected: false,
      resolve: function resolve() {},
      timer: undefined,
      timeoutDuration: meeting.config.reconnection.iceReconnectionTimeout
    };
    /**
     * @instance
     * @type {String}
     * @private
     * @memberof ReconnectionManager
    */

    this.status = _constants.RECONNECTION.STATE.DEFAULT_STATUS;
    /**
     * @instance
     * @type {Number}
     * @private
     * @memberof ReconnectionManager
    */

    this.tryCount = _constants.RECONNECTION.STATE.DEFAULT_TRY_COUNT;
    /**
     * @instance
     * @type {Object}
     * @private
     * @memberof ReconnectionManager
    */

    this.webex = meeting.webex;
    /**
     * @instance
     * @type {Meeting}
     * @private
     * @memberof ReconnectionManager
     */

    this.meeting = meeting;
    this.maxRejoinAttempts = meeting.config.reconnection.maxRejoinAttempts;
    this.rejoinAttempts = _constants.RECONNECTION.STATE.DEFAULT_TRY_COUNT; // Make sure reconnection state is in default

    this.reset();
  }
  /**
   * Sets the iceState to connected and clears any disconnect timeouts and
   * related timeout data within the iceState.
   *
   * @returns {undefined}
   * @public
   * @memberof ReconnectionManager
   */


  (0, _createClass2.default)(ReconnectionManager, [{
    key: "iceReconnected",
    value: function iceReconnected() {
      if (this.iceState.disconnected) {
        _loggerProxy.default.logger.log('ReconnectionManager:index#iceReconnected --> ice has reconnected');

        this.iceState.resolve();

        this.iceState.resolve = function () {};

        if (this.iceState.timer) {
          clearTimeout(this.iceState.timer);
          delete this.iceState.timer;
        }

        this.iceState.disconnected = false;
      }
    }
    /**
     * Set the iceState to disconnected and generates a timeout that waits for the
     * iceState to reconnect and then resolves. If the ice state is already
     * processing a reconnect, it immediately resolves. Rejects if the timeout
     * duration is reached.
     *
     * @returns {Promise<undefined>}
     * @public
     * @memberof ReconnectionManager
     */

  }, {
    key: "waitForIceReconnect",
    value: function waitForIceReconnect() {
      var _this = this;

      if (!this.iceState.disconnected) {
        _loggerProxy.default.logger.log('ReconnectionManager:index#waitForIceReconnect --> waiting for ice reconnect');

        this.iceState.disconnected = true;
        return new _promise.default(function (resolve, reject) {
          _this.iceState.timer = setTimeout(function () {
            if (_this.iceState.disconnected === false) {
              resolve();
            } else {
              _this.iceState.disconnected = false;
              reject(new Error("ice reconnection did not occur in ".concat(_this.iceState.timeoutDuration, "ms")));
            }
          }, _this.iceState.timeoutDuration);
          _this.iceState.resolve = resolve;
        });
      } // return a resolved promise to prevent multiple catch executions of reconnect


      return _promise.default.resolve();
    }
    /**
     * @returns {undefined}
     * @public
     * @memberof ReconnectionManager
     */

  }, {
    key: "reset",
    value: function reset() {
      this.status = _constants.RECONNECTION.STATE.DEFAULT_STATUS;
      this.tryCount = _constants.RECONNECTION.STATE.DEFAULT_TRY_COUNT;
      this.rejoinAttempts = _constants.RECONNECTION.STATE.DEFAULT_TRY_COUNT;
    }
    /**
     * @returns {Boolean}
     * @throws {ReconnectionError}
     * @private
     * @memberof ReconnectionManager
     */

  }, {
    key: "validate",
    value: function validate() {
      if (this.meeting.config.reconnection.enabled) {
        if (this.status === _constants.RECONNECTION.STATE.DEFAULT_STATUS || this.status === _constants.RECONNECTION.STATE.COMPLETE) {
          return true;
        }

        _loggerProxy.default.logger.info('ReconnectionManager:index#validate --> Reconnection already in progress.');

        throw new _reconnectionInProgress.default('Reconnection already in progress.');
      }

      _loggerProxy.default.logger.info('ReconnectionManager:index#validate --> Reconnection is not enabled.');

      throw new _reconnection.default('Reconnection is not enabled.');
    }
    /**
     * Initiates a media reconnect for the active meeting
     * @param {Object} reconnectOptions
     * @param {boolean} [reconnectOptions.networkDisconnect=false] indicates if a network disconnect event happened
     * @returns {Promise}
     * @public
     * @memberof ReconnectionManager
     */

  }, {
    key: "reconnect",
    value: function () {
      var _reconnect = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var _this2 = this;

        var _ref,
            _ref$networkDisconnec,
            networkDisconnect,
            _args = arguments;

        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _ref = _args.length > 0 && _args[0] !== undefined ? _args[0] : {}, _ref$networkDisconnec = _ref.networkDisconnect, networkDisconnect = _ref$networkDisconnec === void 0 ? false : _ref$networkDisconnec;

                _loggerProxy.default.logger.info('ReconnectionManager:index#reconnect --> Reconnection start.'); // First, validate that we can reconnect, if not, it will throw an error


                _context.prev = 2;
                this.validate();
                _context.next = 10;
                break;

              case 6:
                _context.prev = 6;
                _context.t0 = _context["catch"](2);

                _loggerProxy.default.logger.info('ReconnectionManager:index#reconnect --> Reconnection unable to begin.', _context.t0);

                throw _context.t0;

              case 10:
                _metrics.default.postEvent({
                  event: _config.eventType.MEDIA_RECONNECTING,
                  meeting: this.meeting
                });

                return _context.abrupt("return", this.executeReconnection({
                  networkDisconnect: networkDisconnect
                }).then(function () {
                  _loggerProxy.default.logger.info('ReconnectionManager:index#reconnect --> Reconnection successful.');

                  _metrics.default.postEvent({
                    event: _config.eventType.MEDIA_RECOVERED,
                    meeting: _this2.meeting,
                    data: {
                      recoveredBy: _config.reconnection.RECOVERED_BY_NEW
                    }
                  });
                }).catch(function (reconnectError) {
                  if (reconnectError instanceof NeedsRetryError) {
                    _loggerProxy.default.logger.info('ReconnectionManager:index#reconnect --> Reconnection not successful, retrying.'); // Reset our reconnect status since we are looping back to the beginning


                    _this2.status = _constants.RECONNECTION.STATE.DEFAULT_STATUS;
                    return _this2.reconnect({
                      networkDisconnect: true
                    });
                  }

                  _loggerProxy.default.logger.error('ReconnectionManager:index#reconnect --> Reconnection failed.', reconnectError);

                  _metrics.default.postEvent({
                    event: _config.eventType.CALL_ABORTED,
                    meeting: _this2.meeting,
                    // TODO: need to move this to error handler module, create a new one
                    data: {
                      errors: [{
                        category: _config.errorObjects.category.media,
                        errorCode: 2008,
                        fatal: true,
                        name: _config.errorObjects.name.mediaEngine,
                        shownToUser: false
                      }]
                    }
                  });

                  throw reconnectError;
                }));

              case 12:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 6]]);
      }));

      function reconnect() {
        return _reconnect.apply(this, arguments);
      }

      return reconnect;
    }()
    /**
     * @param {Object} reconnectOptions
     * @param {boolean} [reconnectOptions.networkDisconnect=false] indicates if a network disconnect event happened
     * @returns {Promise}
     * @throws {NeedsRetryError}
     * @private
     * @memberof ReconnectionManager
     */

  }, {
    key: "executeReconnection",
    value: function () {
      var _executeReconnection = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(_ref2) {
        var _ref2$networkDisconne, networkDisconnect, rejoinMeeting, wasSharing, media;

        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _ref2$networkDisconne = _ref2.networkDisconnect, networkDisconnect = _ref2$networkDisconne === void 0 ? false : _ref2$networkDisconne;
                this.status = _constants.RECONNECTION.STATE.IN_PROGRESS;

                _loggerProxy.default.logger.info('ReconnectionManager:index#executeReconnection --> Attempting to reconnect to meeting.');

                if (!networkDisconnect) {
                  _context2.next = 14;
                  break;
                }

                _context2.prev = 4;
                _context2.next = 7;
                return this.reconnectMercuryWebSocket();

              case 7:
                _context2.next = 14;
                break;

              case 9:
                _context2.prev = 9;
                _context2.t0 = _context2["catch"](4);

                // TODO: Add Metrics for Reconnect Failed
                _loggerProxy.default.logger.error('ReconnectionManager:index#executeReconnection --> Unable to reconnect to websocket, giving up.');

                this.status = _constants.RECONNECTION.STATE.FAILURE;
                throw _context2.t0;

              case 14:
                rejoinMeeting = false;
                wasSharing = this.meeting.isSharing;
                _context2.prev = 16;

                _loggerProxy.default.logger.info('ReconnectionManager:index#executeReconnection --> Updating meeting data from server.');

                _context2.next = 20;
                return this.webex.meetings.syncMeetings();

              case 20:
                _context2.next = 26;
                break;

              case 22:
                _context2.prev = 22;
                _context2.t1 = _context2["catch"](16);

                _loggerProxy.default.logger.info('ReconnectionManager:index#executeReconnection --> Unable to sync meetings, reconnecting.', _context2.t1);

                throw new NeedsRetryError(_context2.t1);

              case 26:
                if (!(this.meeting.state === _constants._LEFT_)) {
                  _context2.next = 44;
                  break;
                }

                _context2.prev = 27;
                rejoinMeeting = true;
                _context2.next = 31;
                return this.meeting.join();

              case 31:
                _context2.next = 44;
                break;

              case 33:
                _context2.prev = 33;
                _context2.t2 = _context2["catch"](27);
                this.rejoinAttempts += 1;

                if (!(this.rejoinAttempts <= this.maxRejoinAttempts)) {
                  _context2.next = 41;
                  break;
                }

                _loggerProxy.default.logger.info('ReconnectionManager:index#executeReconnection --> Unable to rejoin meeting, reconnecting.', _context2.t2);

                throw new NeedsRetryError(_context2.t2);

              case 41:
                _loggerProxy.default.logger.error('ReconnectionManager:index#executeReconnection --> Unable to rejoin meeting after max attempts.', _context2.t2);

                this.status = _constants.RECONNECTION.STATE.FAILURE;
                throw _context2.t2;

              case 44:
                _context2.prev = 44;
                _context2.next = 47;
                return this.reconnectMedia();

              case 47:
                media = _context2.sent;

                _loggerProxy.default.logger.log('ReconnectionManager:index#executeReconnection --> Media reestablished');

                this.status = _constants.RECONNECTION.STATE.COMPLETE;

                if (rejoinMeeting && wasSharing) {
                  // Stop the share streams if user tried to rejoin
                  _media.default.stopTracks(this.meeting.mediaProperties.shareTrack);

                  this.meeting.isSharing = false;
                  this.meeting.mediaProperties.mediaDirection.sendShare = false;

                  _triggerProxy.default.trigger(this.meeting, {
                    file: 'reconnection-manager/index',
                    function: 'executeReconnection'
                  }, _constants.EVENT_TRIGGERS.MEETING_STOPPED_SHARING_LOCAL, {
                    reason: _constants.SHARE_STOPPED_REASON.MEETING_REJOIN
                  });
                }

                return _context2.abrupt("return", media);

              case 54:
                _context2.prev = 54;
                _context2.t3 = _context2["catch"](44);

                // TODO: Add Metrics for Reconnect Failed
                _loggerProxy.default.logger.error('ReconnectionManager:index#executeReconnection --> Media reestablishment failed');

                this.status = _constants.RECONNECTION.STATE.FAILURE;
                throw _context2.t3;

              case 59:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this, [[4, 9], [16, 22], [27, 33], [44, 54]]);
      }));

      function executeReconnection(_x) {
        return _executeReconnection.apply(this, arguments);
      }

      return executeReconnection;
    }()
    /**
     * @param {Meeting} meeting
     * @returns {Promise}
     * @private
     * @memberof ReconnectionManager
     */

  }, {
    key: "reconnectMedia",
    value: function reconnectMedia() {
      var _this3 = this;

      _loggerProxy.default.logger.log('ReconnectionManager:index#reconnectMedia --> Begin reestablishment of media');

      ReconnectionManager.setupPeerConnection(this.meeting);
      return _media.default.attachMedia(this.meeting.mediaProperties, {
        meetingId: this.meeting.id,
        remoteQualityLevel: this.meeting.mediaProperties.remoteQualityLevel,
        enableRtx: this.meeting.config.enableRtx
      }).then(function (peerConnection) {
        return _this3.meeting.setRemoteStream(peerConnection);
      }).then(function () {
        return _this3.meeting.roap.sendRoapMediaRequest({
          sdp: _this3.meeting.mediaProperties.peerConnection.sdp,
          roapSeq: _this3.meeting.roapSeq,
          meeting: _this3.meeting,
          reconnect: true // Need to check if its a reconnect after rejoin or media inactivity

        });
      });
    }
    /**
     * Attempt to Reconnect Mercury Websocket
     * @returns {Promise}
     * @private
     * @memberof ReconnectionManager
     */

  }, {
    key: "reconnectMercuryWebSocket",
    value: function () {
      var _reconnectMercuryWebSocket = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee3() {
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _loggerProxy.default.logger.info('ReconnectionManager:index#reconnectMercuryWebSocket --> Reconnecting websocket.'); // First, attempt to disconnect if we think we are already connected.


                if (!this.webex.internal.mercury.connected) {
                  _context3.next = 13;
                  break;
                }

                _loggerProxy.default.logger.info('ReconnectionManager:index#reconnectMercuryWebSocket --> Disconnecting existing websocket.');

                _context3.prev = 3;
                _context3.next = 6;
                return this.webex.internal.mercury.disconnect();

              case 6:
                _loggerProxy.default.logger.info('ReconnectionManager:index#reconnectMercuryWebSocket --> Websocket disconnected successfully.');

                _context3.next = 13;
                break;

              case 9:
                _context3.prev = 9;
                _context3.t0 = _context3["catch"](3);

                // If we can't disconnect, the sdk is in such a bad state that reconnecting is not going to happen.
                _loggerProxy.default.logger.error('ReconnectionManager:index#reconnectMercuryWebSocket --> Unable to disconnect to websocket, giving up.', _context3.t0);

                throw _context3.t0;

              case 13:
                _context3.prev = 13;

                _loggerProxy.default.logger.info('ReconnectionManager:index#reconnectMercuryWebSocket --> Connecting websocket.');

                _context3.next = 17;
                return this.webex.internal.mercury.connect();

              case 17:
                _loggerProxy.default.logger.info('ReconnectionManager:index#reconnectMercuryWebSocket --> Websocket connected successfully.');

                _context3.next = 24;
                break;

              case 20:
                _context3.prev = 20;
                _context3.t1 = _context3["catch"](13);

                _loggerProxy.default.logger.error('ReconnectionManager:index#reconnectMercuryWebSocket --> Unable to connect to websocket, giving up.', _context3.t1);

                throw _context3.t1;

              case 24:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this, [[3, 9], [13, 20]]);
      }));

      function reconnectMercuryWebSocket() {
        return _reconnectMercuryWebSocket.apply(this, arguments);
      }

      return reconnectMercuryWebSocket;
    }()
    /**
     * @param {Meeting} meeting
     * @returns {undefined}
     * @private
     * @memberof ReconnectionManager
     */

  }], [{
    key: "setupPeerConnection",
    value: function setupPeerConnection(meeting) {
      // close pcs, unset to null and create a new one with out closing any streams
      _peerConnectionManager.default.close(meeting.mediaProperties.peerConnection);

      meeting.mediaProperties.unsetPeerConnection();
      meeting.mediaProperties.reInitiatePeerconnection();

      _peerConnectionManager.default.setPeerConnectionEvents(meeting); // update the peerconnection in the stats manager when ever we reconnect


      meeting.statsAnalyzer.updatePeerconnection(meeting.mediaProperties.peerConnection);
    }
  }]);
  return ReconnectionManager;
}();

exports.default = ReconnectionManager;
